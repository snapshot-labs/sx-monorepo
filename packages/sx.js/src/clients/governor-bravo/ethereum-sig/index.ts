import {
  Signer,
  TypedDataField,
  TypedDataSigner
} from '@ethersproject/abstract-signer';
import { Contract } from '@ethersproject/contracts';
import GovernorAbi from './abis/Governor.json';
import { ballotTypes, ballotWithReasonTypes } from './types';
import { GovernorBravoAuthenticator } from '../../../types';
import { EIP712Ballot, Envelope, SignatureData, Vote } from '../types';

const CUSTOM_NAMES: Record<string, string | undefined> = {
  // This is our Sepolia test deployment
  // Looks like I deployed some weird version that doesn't have name function
  '0x69112D158A607DD388034c0C09242FF966985258': 'Compound Governor Bravo'
};

export class EthereumSig {
  chainId: number;

  constructor({ chainId }: { chainId: number }) {
    this.chainId = chainId;
  }

  public async sign<T extends EIP712Ballot>(
    signer: Signer & TypedDataSigner,
    authenticatorType: GovernorBravoAuthenticator,
    verifyingContract: string,
    message: T,
    types: Record<string, TypedDataField[]>
  ): Promise<SignatureData> {
    const address = await signer.getAddress();

    const govenorContract = new Contract(
      verifyingContract,
      GovernorAbi,
      signer
    );

    const name =
      CUSTOM_NAMES[verifyingContract] ?? (await govenorContract.name());

    const domain = {
      name,
      chainId: this.chainId,
      verifyingContract
    };

    const signature = await signer._signTypedData(domain, types, message);

    return {
      authenticatorType,
      address,
      signature,
      domain,
      types,
      message
    };
  }

  async vote({
    signer,
    authenticatorType,
    data
  }: {
    signer: Signer & TypedDataSigner;
    authenticatorType: GovernorBravoAuthenticator;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const { spaceId, proposalId, choice: support, reason } = data;

    let message;
    if (reason) {
      message = {
        proposalId,
        support,
        reason
      };
    } else {
      message = {
        proposalId,
        support
      };
    }

    const signatureData = await this.sign(
      signer,
      authenticatorType,
      spaceId,
      message,
      reason ? ballotWithReasonTypes : ballotTypes
    );

    return {
      data,
      signatureData
    };
  }
}
