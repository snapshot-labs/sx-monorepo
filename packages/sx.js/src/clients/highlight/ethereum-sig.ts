import {
  Signer,
  TypedDataDomain,
  TypedDataField,
  TypedDataSigner
} from '@ethersproject/abstract-signer';
import {
  CloseDiscussion,
  CreateDiscussion,
  CreateStatement,
  Envelope,
  HideStatement,
  PinStatement,
  SetAlias,
  UnpinStatement,
  Vote
} from './types';
import {
  ALIASES_CONFIG,
  HIGHLIGHT_DOMAIN,
  TOWNHALL_CONFIG
} from '../../highlightConstants';

export class HighlightEthereumSigClient {
  private highlightUrl: string;

  constructor(highlightUrl: string) {
    this.highlightUrl = highlightUrl;
  }

  public async getDomain(
    signer: Signer & TypedDataSigner,
    salt: bigint,
    to: string
  ): Promise<Required<TypedDataDomain>> {
    const chainId = await signer.getChainId();

    const domain = {
      ...HIGHLIGHT_DOMAIN,
      chainId,
      salt: `0x${salt.toString(16).padStart(64, '0')}`,
      verifyingContract: to
    };

    return domain;
  }

  public async sign(
    signer: Signer & TypedDataSigner,
    domain: Required<TypedDataDomain>,
    types: Record<string, TypedDataField[]>,
    message: Record<string, any>
  ) {
    return signer._signTypedData(domain, types, message);
  }

  public async send(envelope: Envelope) {
    const { domain, message, primaryType, signer, signature } = envelope;

    const payload = {
      method: 'hl_postMessage',
      params: {
        domain: {
          ...domain,
          salt: domain.salt.toString()
        },
        message,
        primaryType,
        signer,
        signature
      }
    };

    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    };

    const res = await fetch(this.highlightUrl, body);

    if (!res.ok) {
      throw new Error(`Failed to send envelope: ${res.statusText}`);
    }

    const result = await res.json();

    if (result.error) {
      throw new Error(result.error);
    }

    return result;
  }

  public async setAlias({
    signer,
    data,
    salt
  }: {
    signer: Signer & TypedDataSigner;
    data: SetAlias;
    salt: bigint;
  }): Promise<Envelope> {
    const domain = await this.getDomain(signer, salt, ALIASES_CONFIG.address);

    const { alias, from } = data;
    const message = {
      from,
      alias
    };

    const signature = await this.sign(
      signer,
      domain,
      ALIASES_CONFIG.types.setAlias,
      message
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain,
      message,
      primaryType: 'SetAlias',
      signer: await signer.getAddress(),
      signature
    };
  }

  public async createDiscussion({
    signer,
    data,
    salt
  }: {
    signer: Signer & TypedDataSigner;
    data: CreateDiscussion;
    salt: bigint;
  }): Promise<Envelope> {
    const domain = await this.getDomain(signer, salt, TOWNHALL_CONFIG.address);

    const { title, body, discussionUrl } = data;
    const message = {
      title,
      body,
      discussionUrl
    };

    const signature = await this.sign(
      signer,
      domain,
      TOWNHALL_CONFIG.types.createDiscussion,
      message
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain,
      message,
      primaryType: 'Discussion',
      signer: await signer.getAddress(),
      signature
    };
  }

  public async closeDiscussion({
    signer,
    data,
    salt
  }: {
    signer: Signer & TypedDataSigner;
    data: CloseDiscussion;
    salt: bigint;
  }): Promise<Envelope> {
    const domain = await this.getDomain(signer, salt, TOWNHALL_CONFIG.address);

    const { discussion } = data;
    const message = {
      discussion
    };

    const signature = await this.sign(
      signer,
      domain,
      TOWNHALL_CONFIG.types.closeDiscussion,
      message
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain,
      message,
      primaryType: 'CloseDiscussion',
      signer: await signer.getAddress(),
      signature
    };
  }

  public async createStatement({
    signer,
    data,
    salt
  }: {
    signer: Signer & TypedDataSigner;
    data: CreateStatement;
    salt: bigint;
  }): Promise<Envelope> {
    const domain = await this.getDomain(signer, salt, TOWNHALL_CONFIG.address);

    const { discussion, statement } = data;
    const message = {
      discussion,
      statement
    };

    const signature = await this.sign(
      signer,
      domain,
      TOWNHALL_CONFIG.types.createStatement,
      message
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain,
      message,
      primaryType: 'Statement',
      signer: await signer.getAddress(),
      signature
    };
  }

  public async hideStatement({
    signer,
    data,
    salt
  }: {
    signer: Signer & TypedDataSigner;
    data: HideStatement;
    salt: bigint;
  }): Promise<Envelope> {
    const domain = await this.getDomain(signer, salt, TOWNHALL_CONFIG.address);

    const { discussion, statement } = data;
    const message = {
      discussion,
      statement
    };

    const signature = await this.sign(
      signer,
      domain,
      TOWNHALL_CONFIG.types.hideStatement,
      message
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain,
      message,
      primaryType: 'HideStatement',
      signer: await signer.getAddress(),
      signature
    };
  }

  public async pinStatement({
    signer,
    data,
    salt
  }: {
    signer: Signer & TypedDataSigner;
    data: PinStatement;
    salt: bigint;
  }): Promise<Envelope> {
    const domain = await this.getDomain(signer, salt, TOWNHALL_CONFIG.address);

    const { discussion, statement } = data;
    const message = {
      discussion,
      statement
    };

    const signature = await this.sign(
      signer,
      domain,
      TOWNHALL_CONFIG.types.pinStatement,
      message
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain,
      message,
      primaryType: 'PinStatement',
      signer: await signer.getAddress(),
      signature
    };
  }

  public async unpinStatement({
    signer,
    data,
    salt
  }: {
    signer: Signer & TypedDataSigner;
    data: UnpinStatement;
    salt: bigint;
  }): Promise<Envelope> {
    const domain = await this.getDomain(signer, salt, TOWNHALL_CONFIG.address);

    const { discussion, statement } = data;
    const message = {
      discussion,
      statement
    };

    const signature = await this.sign(
      signer,
      domain,
      TOWNHALL_CONFIG.types.unpinStatement,
      message
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain,
      message,
      primaryType: 'UnpinStatement',
      signer: await signer.getAddress(),
      signature
    };
  }

  public async vote({
    signer,
    data,
    salt
  }: {
    signer: Signer & TypedDataSigner;
    data: Vote;
    salt: bigint;
  }): Promise<Envelope> {
    const domain = await this.getDomain(signer, salt, TOWNHALL_CONFIG.address);

    const { discussion, statement, choice } = data;
    const message = {
      discussion,
      statement,
      choice
    };

    const signature = await this.sign(
      signer,
      domain,
      TOWNHALL_CONFIG.types.vote,
      message
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain,
      message,
      primaryType: 'Vote',
      signer: await signer.getAddress(),
      signature
    };
  }
}
