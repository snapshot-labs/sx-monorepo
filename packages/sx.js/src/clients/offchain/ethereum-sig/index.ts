import {
  Signer,
  TypedDataField,
  TypedDataSigner
} from '@ethersproject/abstract-signer';
import {
  aliasTypes,
  approvalVoteTypes,
  basicVoteTypes,
  cancelProposalTypes,
  deleteSpaceTypes,
  domain,
  encryptedVoteTypes,
  followSpaceTypes,
  proposeTypes,
  rankedChoiceVoteTypes,
  singleChoiceVoteTypes,
  unfollowSpaceTypes,
  updateProposalTypes,
  updateSpaceTypes,
  updateStatementTypes,
  updateUserTypes,
  weightedVoteTypes
} from './types';
import { offchainGoerli } from '../../../offchainNetworks';
import { OffchainNetworkConfig, SignatureData } from '../../../types';
import {
  CancelProposal,
  DeleteSpace,
  EIP712CancelProposalMessage,
  EIP712DeleteSpaceMessage,
  EIP712FollowSpaceMessage,
  EIP712Message,
  EIP712ProposeMessage,
  EIP712SetAliasMessage,
  EIP712UnfollowSpaceMessage,
  EIP712UpdateProposal,
  EIP712UpdateSpaceMessage,
  EIP712UpdateStatementMessage,
  EIP712UpdateUserMessage,
  EIP712VoteMessage,
  Envelope,
  FollowSpace,
  Propose,
  SetAlias,
  UnfollowSpace,
  UpdateProposal,
  UpdateSpace,
  UpdateStatement,
  UpdateUser,
  Vote
} from '../types';
import { encryptChoices } from '../utils';

const SEQUENCER_URLS: Record<OffchainNetworkConfig['eip712ChainId'], string> = {
  1: 'https://seq.snapshot.org',
  5: 'https://testnet.seq.snapshot.org'
};

const RELAYER_URLS: Record<OffchainNetworkConfig['eip712ChainId'], string> = {
  1: 'https://relayer.snapshot.org',
  5: 'https://testnet.seq.snapshot.org' // no relayer for testnet
};

type EthereumSigClientOpts = {
  networkConfig?: OffchainNetworkConfig;
  sequencerUrl?: string;
};

export class EthereumSig {
  sequencerUrl: string;
  networkConfig: OffchainNetworkConfig;

  constructor(opts?: EthereumSigClientOpts) {
    this.networkConfig = opts?.networkConfig || offchainGoerli;
    this.sequencerUrl =
      opts?.sequencerUrl || SEQUENCER_URLS[this.networkConfig.eip712ChainId];
  }

  public async sign<
    T extends
      | EIP712VoteMessage
      | EIP712ProposeMessage
      | EIP712UpdateProposal
      | EIP712CancelProposalMessage
      | EIP712FollowSpaceMessage
      | EIP712UnfollowSpaceMessage
      | EIP712SetAliasMessage
      | EIP712UpdateUserMessage
      | EIP712UpdateStatementMessage
      | EIP712UpdateSpaceMessage
      | EIP712DeleteSpaceMessage
  >(
    signer: Signer & TypedDataSigner,
    message: T,
    types: Record<string, TypedDataField[]>
  ): Promise<SignatureData> {
    const address = await signer.getAddress();
    const EIP712Message: EIP712Message = {
      from: address,
      timestamp: parseInt((Date.now() / 1000).toFixed()),
      ...message
    };
    const signature = await signer._signTypedData(domain, types, EIP712Message);
    return {
      address,
      signature,
      domain,
      types,
      message: EIP712Message
    };
  }

  public async send(
    envelope: Envelope<
      | Vote
      | Propose
      | UpdateProposal
      | CancelProposal
      | FollowSpace
      | UnfollowSpace
      | SetAlias
    >
  ) {
    const {
      address,
      signature: sig,
      domain,
      types,
      primaryType,
      message
    } = envelope.signatureData!;
    const payload: any = {
      address,
      sig,
      data: {
        domain,
        types,
        message
      }
    };

    // primaryType needs to be attached when sending starknet-sig generated payload
    if (primaryType) payload.data.primaryType = primaryType;

    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    };

    let url = this.sequencerUrl;
    if (sig === '0x') {
      url = RELAYER_URLS[this.networkConfig.eip712ChainId] || url;
    }
    const res = await fetch(url, body);
    const result = await res.json();

    if (result.error) {
      throw new Error(
        typeof result.error_description === 'string'
          ? result.error_description
          : result.error
      );
    }

    return result;
  }

  public async propose({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const signatureData = await this.sign(signer, data, proposeTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const signatureData = await this.sign(signer, data, updateProposalTypes);

    return {
      signatureData,
      data
    };
  }

  public async cancel({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: CancelProposal;
  }): Promise<Envelope<CancelProposal>> {
    const signatureData = await this.sign(signer, data, cancelProposalTypes);

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    let choice: EIP712VoteMessage['choice'];
    let voteType: { Vote: { name: string; type: string }[] };

    switch (data.type) {
      case 'single-choice':
        voteType = singleChoiceVoteTypes;
        choice = data.choice as number;
        break;
      case 'approval':
        voteType = approvalVoteTypes;
        choice = data.choice as number[];
        break;
      case 'ranked-choice':
        voteType = rankedChoiceVoteTypes;
        choice = data.choice as number[];
        break;
      case 'weighted':
      case 'quadratic':
        voteType = weightedVoteTypes;
        choice = JSON.stringify(data.choice);
        break;
      default:
        voteType = basicVoteTypes;
        choice = data.choice as number;
    }

    const message: EIP712VoteMessage = {
      space: data.space,
      proposal: data.proposal,
      choice,
      reason: data.reason || '',
      app: data.app,
      metadata: ''
    };

    if (data.privacy) {
      message.privacy = data.privacy;
      voteType = encryptedVoteTypes;
      message.choice = await encryptChoices(
        data.privacy,
        data.proposal,
        typeof message.choice === 'string'
          ? message.choice
          : JSON.stringify(message.choice)
      );
    }
    const signatureData = await this.sign(signer, message, voteType);

    return {
      signatureData,
      data
    };
  }

  public async followSpace({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: FollowSpace;
  }): Promise<Envelope<FollowSpace>> {
    const signatureData = await this.sign(signer, data, followSpaceTypes);

    return {
      signatureData,
      data
    };
  }

  public async unfollowSpace({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UnfollowSpace;
  }): Promise<Envelope<UnfollowSpace>> {
    const signatureData = await this.sign(signer, data, unfollowSpaceTypes);

    return {
      signatureData,
      data
    };
  }

  public async setAlias({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: SetAlias;
  }): Promise<Envelope<SetAlias>> {
    const signatureData = await this.sign(signer, data, aliasTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateUser({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateUser;
  }) {
    const signatureData = await this.sign(signer, data, updateUserTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateStatement({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateStatement;
  }) {
    const signatureData = await this.sign(signer, data, updateStatementTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateSpace({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateSpace;
  }) {
    const signatureData = await this.sign(signer, data, updateSpaceTypes);

    return {
      signatureData,
      data
    };
  }

  public async deleteSpace({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: DeleteSpace;
  }) {
    const signatureData = await this.sign(signer, data, deleteSpaceTypes);

    return {
      signatureData,
      data
    };
  }
}
