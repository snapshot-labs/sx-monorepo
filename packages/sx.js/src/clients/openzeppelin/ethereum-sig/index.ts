import {
  Signer,
  TypedDataField,
  TypedDataSigner
} from '@ethersproject/abstract-signer';
import { BigNumber } from '@ethersproject/bignumber';
import { Contract } from '@ethersproject/contracts';
import GovernorV4Abi from './abis/GovernorV4.json';
import GovernorV5Abi from './abis/GovernorV5.json';
import { ballotTypesV5, extendedBallotTypesV5 } from './types';
import { OpenZeppelinAuthenticator } from '../../../types';
import {
  EIP712BallotV4,
  EIP712BallotV5,
  EIP712ExtendedBallotV5,
  Envelope,
  SignatureData,
  Vote
} from '../types';

export class EthereumSig {
  chainId: number;

  constructor({ chainId }: { chainId: number }) {
    this.chainId = chainId;
  }

  public async sign<
    T extends EIP712BallotV4 | EIP712BallotV5 | EIP712ExtendedBallotV5
  >(
    signer: Signer & TypedDataSigner,
    authenticatorType: OpenZeppelinAuthenticator,
    verifyingContract: string,
    message: T,
    types: Record<string, TypedDataField[]>
  ): Promise<SignatureData> {
    const address = await signer.getAddress();

    const govenorContract = new Contract(
      verifyingContract,
      GovernorV4Abi,
      signer
    );

    const [name, version]: [string, string] = await Promise.all([
      govenorContract.name(),
      govenorContract.version()
    ]);

    const domain = {
      name,
      version,
      chainId: this.chainId,
      verifyingContract
    };

    const signature = await signer._signTypedData(domain, types, message);

    return {
      authenticatorType,
      address,
      signature,
      domain,
      types,
      message
    };
  }

  async vote({
    signer,
    authenticatorType,
    data
  }: {
    signer: Signer & TypedDataSigner;
    authenticatorType: OpenZeppelinAuthenticator;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const voter = await signer.getAddress();
    const { spaceId, proposalId, choice: support, reason } = data;

    let message: EIP712BallotV4 | EIP712BallotV5 | EIP712ExtendedBallotV5;

    if (authenticatorType === 'OpenZeppelinAuthenticatorSignatureV5') {
      const govenorContract = new Contract(spaceId, GovernorV5Abi, signer);
      const nonce: BigNumber = await govenorContract.nonces(voter);

      if (reason) {
        message = {
          voter,
          proposalId,
          support,
          nonce: nonce.toNumber(),
          reason,
          params: '0x'
        };
      } else {
        message = {
          voter,
          proposalId,
          support,
          nonce: nonce.toNumber()
        };
      }
    } else if (authenticatorType === 'OpenZeppelinAuthenticatorSignatureV4') {
      message = {
        proposalId,
        support
      };
    } else {
      throw new Error('Unsupported authenticator type');
    }

    const signatureData = await this.sign(
      signer,
      authenticatorType,
      spaceId,
      message,
      reason ? extendedBallotTypesV5 : ballotTypesV5
    );

    return {
      data,
      signatureData
    };
  }
}
