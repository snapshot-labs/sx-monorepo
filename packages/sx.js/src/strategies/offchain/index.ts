import createOnlyMembersStrategy from './only-members';
import createRemoteValidateStrategy from './remote-validate';
import createRemoteVpStrategy from './remote-vp';
import { Strategy } from '../../clients/offchain/types';

export function getStrategy(name: string): Strategy | null {
  if (name === 'only-members') return createOnlyMembersStrategy();

  if (['any', 'basic', 'passport-gated', 'arbitrum', 'karma-eas-attestation'].includes(name)) {
    return createRemoteValidateStrategy(name);
  }

  return createRemoteVpStrategy();
}
