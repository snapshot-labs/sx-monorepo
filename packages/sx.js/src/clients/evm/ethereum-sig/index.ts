import { AbiCoder } from '@ethersproject/abi';
import {
  Signer,
  TypedDataField,
  TypedDataSigner
} from '@ethersproject/abstract-signer';
import randomBytes from 'randombytes';
import {
  domain as baseDomain,
  proposeTypes,
  updateProposalTypes,
  voteTypes
} from './types';
import { getStrategiesWithParams } from '../../../strategies/evm';
import { bytesToHex } from '../../../utils/bytes';
import { SplitUint256 } from '../../../utils/split-uint256';
import {
  ClientConfig,
  ClientOpts,
  EIP712ProposeMessage,
  EIP712UpdateProposalMessage,
  EIP712VoteMessage,
  Envelope,
  Propose,
  SignatureData,
  UpdateProposal,
  Vote
} from '../types';

export class EthereumSig {
  config: ClientConfig & { manaUrl: string };

  constructor(opts: ClientOpts & { manaUrl: string }) {
    this.config = opts;
  }

  generateSalt() {
    return Number(SplitUint256.fromHex(bytesToHex(randomBytes(4))).toHex());
  }

  public async sign<
    T extends
      | EIP712ProposeMessage
      | EIP712UpdateProposalMessage
      | EIP712VoteMessage
  >(
    signer: Signer & TypedDataSigner,
    verifyingContract: string,
    message: T,
    types: Record<string, TypedDataField[]>
  ): Promise<SignatureData> {
    const address = await signer.getAddress();

    const domain = {
      ...baseDomain,
      chainId: this.config.networkConfig.eip712ChainId,
      verifyingContract
    };

    const signature = await signer._signTypedData(domain, types, message);

    return {
      address,
      signature,
      domain,
      types,
      message
    };
  }

  public async send(envelope: Envelope<Propose | UpdateProposal | Vote>) {
    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelope },
        id: null
      })
    };

    const res = await fetch(
      `${this.config.manaUrl}/eth_rpc/${this.config.networkConfig.eip712ChainId}`,
      body
    );
    const json = await res.json();

    return json.result;
  }

  public async propose({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const author = await signer.getAddress();

    const userStrategies = await getStrategiesWithParams(
      'propose',
      data.strategies,
      author,
      data,
      this.config
    );

    const abiCoder = new AbiCoder();
    const message: EIP712ProposeMessage = {
      space: data.space,
      author,
      metadataURI: data.metadataUri,
      executionStrategy: data.executionStrategy,
      userProposalValidationParams: abiCoder.encode(
        ['tuple(int8 index, bytes params)[]'],
        [userStrategies]
      ),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      proposeTypes
    );

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const author = await signer.getAddress();

    const message: EIP712UpdateProposalMessage = {
      space: data.space,
      author,
      proposalId: data.proposal,
      executionStrategy: data.executionStrategy,
      metadataURI: data.metadataUri,
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      updateProposalTypes
    );

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const voter = await signer.getAddress();

    const userVotingStrategies = await getStrategiesWithParams(
      'vote',
      data.strategies,
      voter,
      data,
      this.config
    );

    const message: EIP712VoteMessage = {
      space: data.space,
      voter,
      proposalId: data.proposal,
      choice: data.choice,
      userVotingStrategies,
      voteMetadataURI: data.metadataUri
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      voteTypes
    );

    return {
      signatureData,
      data
    };
  }
}
