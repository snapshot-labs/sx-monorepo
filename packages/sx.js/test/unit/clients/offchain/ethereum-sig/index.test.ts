import { Wallet } from '@ethersproject/wallet';
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest';
import { EthereumSig } from '../../../../../src/clients/offchain/ethereum-sig';
import { offchainGoerli } from '../../../../../src/offchainNetworks';

describe('EthereumSig', () => {
  // Test address: 0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90
  const TEST_PK =
    'ef4bcf36b5d026b703b86a311031fe2291b979620f01443f795fa213f9105e35';
  const signer = new Wallet(TEST_PK);
  const client = new EthereumSig({ networkConfig: offchainGoerli });

  beforeAll(() => {
    vi.useFakeTimers({
      now: new Date('2024-01-21').getTime()
    });
  });

  afterAll(() => {
    vi.restoreAllMocks();
  });

  it('should create vote envelope', async () => {
    const payload = {
      space: 'wan-test.eth',
      authenticator: '',
      strategies: [],
      proposal:
        '0xcc47146e5e0ac781e8976405a8da4468f2a0c4cdf0c7659353d728fafe46f801',
      choice: 1,
      metadataUri: '',
      type: 'basic'
    };

    const envelope = await client.vote({
      signer,
      data: payload
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create propose envelope', async () => {
    const payload = {
      space: 'wan-test.eth',
      title: 'Creation test',
      body: 'Dummy body',
      type: 'basic',
      discussion: 'https://snapshot.org',
      choices: ['For', 'Against', 'Abstain'],
      labels: ['1234e'],
      start: Math.floor(Date.now() / 1000),
      end: Math.floor(Date.now() / 1000) + 60 * 60,
      snapshot: 19283932,
      plugins: '{}',
      app: 'snapshot-v2'
    };

    const envelope = await client.propose({
      signer,
      data: payload
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create cancelProposal envelope', async () => {
    const payload = {
      space: 'test.eth',
      proposal:
        '0x56b857b02d573b0ba747333b57cb3dd11df57cc0d1bcc41c3c990466b477c5e8'
    };

    const envelope = await client.cancel({
      signer,
      data: payload
    });

    expect(envelope).toMatchSnapshot();
  });
});
