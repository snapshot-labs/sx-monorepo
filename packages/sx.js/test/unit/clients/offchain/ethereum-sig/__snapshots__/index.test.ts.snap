// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EthereumSig > should create cancelProposal envelope 1`] = `
{
  "data": {
    "proposal": "0x56b857b02d573b0ba747333b57cb3dd11df57cc0d1bcc41c3c990466b477c5e8",
    "space": "test.eth",
  },
  "signatureData": {
    "address": "0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90",
    "domain": {
      "name": "snapshot",
      "version": "0.1.4",
    },
    "message": {
      "from": "0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90",
      "proposal": "0x56b857b02d573b0ba747333b57cb3dd11df57cc0d1bcc41c3c990466b477c5e8",
      "space": "test.eth",
      "timestamp": 1705795200,
    },
    "signature": "0xe358de8e39c6c0cc9df929892e61806371e6539199a2b0781b9100700760067334024ca8cd5f86f1cb16fc93e76602989618eeacebf1fafbc1acad882d14b4d61c",
    "types": {
      "CancelProposal": [
        {
          "name": "from",
          "type": "address",
        },
        {
          "name": "space",
          "type": "string",
        },
        {
          "name": "timestamp",
          "type": "uint64",
        },
        {
          "name": "proposal",
          "type": "bytes32",
        },
      ],
    },
  },
}
`;

exports[`EthereumSig > should create propose envelope 1`] = `
{
  "data": {
    "app": "snapshot-v2",
    "body": "Dummy body",
    "choices": [
      "For",
      "Against",
      "Abstain",
    ],
    "discussion": "https://snapshot.org",
    "end": 1705798800,
    "labels": [
      "1234e",
    ],
    "plugins": "{}",
    "snapshot": 19283932,
    "space": "wan-test.eth",
    "start": 1705795200,
    "title": "Creation test",
    "type": "basic",
  },
  "signatureData": {
    "address": "0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90",
    "domain": {
      "name": "snapshot",
      "version": "0.1.4",
    },
    "message": {
      "app": "snapshot-v2",
      "body": "Dummy body",
      "choices": [
        "For",
        "Against",
        "Abstain",
      ],
      "discussion": "https://snapshot.org",
      "end": 1705798800,
      "from": "0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90",
      "labels": [
        "1234e",
      ],
      "plugins": "{}",
      "snapshot": 19283932,
      "space": "wan-test.eth",
      "start": 1705795200,
      "timestamp": 1705795200,
      "title": "Creation test",
      "type": "basic",
    },
    "signature": "0x33482a4ebefdcc7c5dc006a8766e1b0dc5c653360b894d49a0c7819d97f16b6b67f4b78a9c4cd46f551fef61e4de97be9140bb32d5a43f7b0e52b4936a08c5f41c",
    "types": {
      "Proposal": [
        {
          "name": "from",
          "type": "address",
        },
        {
          "name": "space",
          "type": "string",
        },
        {
          "name": "timestamp",
          "type": "uint64",
        },
        {
          "name": "type",
          "type": "string",
        },
        {
          "name": "title",
          "type": "string",
        },
        {
          "name": "body",
          "type": "string",
        },
        {
          "name": "discussion",
          "type": "string",
        },
        {
          "name": "choices",
          "type": "string[]",
        },
        {
          "name": "labels",
          "type": "string[]",
        },
        {
          "name": "start",
          "type": "uint64",
        },
        {
          "name": "end",
          "type": "uint64",
        },
        {
          "name": "snapshot",
          "type": "uint64",
        },
        {
          "name": "plugins",
          "type": "string",
        },
        {
          "name": "app",
          "type": "string",
        },
      ],
    },
  },
}
`;

exports[`EthereumSig > should create vote envelope 1`] = `
{
  "data": {
    "app": "snapshot-v2",
    "authenticator": "",
    "choice": 1,
    "metadataUri": "",
    "proposal": "0xcc47146e5e0ac781e8976405a8da4468f2a0c4cdf0c7659353d728fafe46f801",
    "space": "wan-test.eth",
    "strategies": [],
    "type": "basic",
  },
  "signatureData": {
    "address": "0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90",
    "domain": {
      "name": "snapshot",
      "version": "0.1.4",
    },
    "message": {
      "app": "snapshot-v2",
      "choice": 1,
      "from": "0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90",
      "metadata": "",
      "proposal": "0xcc47146e5e0ac781e8976405a8da4468f2a0c4cdf0c7659353d728fafe46f801",
      "reason": "",
      "space": "wan-test.eth",
      "timestamp": 1705795200,
    },
    "signature": "0x92df2e1de71e2f0cdf04d7fd5d028fe0d38b990aa896ad677c4627bfd303cd386b17ad4c403dade15f2508b5fa444d2e8b4b1bed8ca63d341006a4d5d2e64bac1b",
    "types": {
      "Vote": [
        {
          "name": "from",
          "type": "address",
        },
        {
          "name": "space",
          "type": "string",
        },
        {
          "name": "timestamp",
          "type": "uint64",
        },
        {
          "name": "proposal",
          "type": "string",
        },
        {
          "name": "choice",
          "type": "uint32",
        },
        {
          "name": "reason",
          "type": "string",
        },
        {
          "name": "app",
          "type": "string",
        },
        {
          "name": "metadata",
          "type": "string",
        },
      ],
    },
  },
}
`;
